ompi/mca/pml/ob1/pml_ob1_accelerator.c

```diff
@@ -0,0 +1,599 @@
+
+    accelerator_event_htod_array = (opal_accelerator_event_t *) calloc(accelerator_event_max, sizeof(opal_accelerator_event_t));
+    if (NULL == accelerator_event_htod_array) {
+        opal_output_verbose(1, mca_pml_ob1_output, "No memory.");
+        rc = OPAL_ERROR;
+        goto cleanup_and_error;
+    }
+
+    /* Create the events since they can be reused. */
+    for (i = 0; i < accelerator_event_max; i++) {
+        result = opal_accelerator.create_event(&accelerator_event_htod_array[i]);
+        if (OPAL_SUCCESS != result) {
+            opal_output_verbose(1, mca_pml_ob1_output, "Accelerator create event failed.");
+            rc = OPAL_ERROR;
+            goto cleanup_and_error;
+        }
+    }
+
+    /* The first available status index is 0.  Make an empty frag
+       array. */
+    accelerator_event_htod_frag_array = (struct mca_btl_base_descriptor_t **) malloc(
+        sizeof(struct mca_btl_base_descriptor_t *) * accelerator_event_max);
+    if (NULL == accelerator_event_htod_frag_array) {
+        opal_output_verbose(1, mca_pml_ob1_output, "No memory.");
+        rc = OPAL_ERROR;
+        goto cleanup_and_error;
+    }
+
+cleanup_and_error:
+    if (OPAL_SUCCESS != rc) {
+        if (NULL != accelerator_event_dtoh_array) {
+            free(accelerator_event_dtoh_array);
+        }
+        if (NULL != accelerator_event_dtoh_frag_array) {
+            free(accelerator_event_dtoh_frag_array);
+        }
+        if (NULL != accelerator_event_htod_array) {
+            free(accelerator_event_htod_array);
+        }
+        if (NULL != accelerator_event_htod_frag_array) {
+            free(accelerator_event_htod_frag_array);
+        }
+    }
+
+    return rc;
+}
+
+void mca_pml_ob1_accelerator_fini(void)
+{
+    int i;
+
+    if (0 == strcmp(accelerator_base_selected_component.base_version.mca_component_name, "null")) {
+        return;
+    }
+
+    if (NULL != accelerator_event_htod_array) {
+        for (i = 0; i < accelerator_event_max; i++) {
+            if (NULL != &accelerator_event_htod_array[i]) {
+                opal_accelerator.destroy_event(&accelerator_event_htod_array[i]);
+            }
+        }
+        free(accelerator_event_htod_array);
+    }
+
+    if (NULL != accelerator_event_dtoh_array) {
+        for (i = 0; i < accelerator_event_max; i++) {
+            if (NULL != &accelerator_event_dtoh_array[i]) {
+                opal_accelerator.destroy_event(&accelerator_event_dtoh_array[i]);
+            }
+        }
+        free(accelerator_event_dtoh_array);
+    }
+    if (NULL != accelerator_event_dtoh_frag_array) {
+        free(accelerator_event_dtoh_frag_array);
+    }
+    if (NULL != accelerator_event_htod_frag_array) {
+        free(accelerator_event_htod_frag_array);
+    }
+
+    opal_accelerator.destroy_stream(&htod_stream);
+    opal_accelerator.destroy_stream(&dtoh_stream);
+
+    OBJ_DESTRUCT(&pml_ob1_accelerator_htod_lock);
+    OBJ_DESTRUCT(&pml_ob1_accelerator_dtoh_lock);
+}
```

Comment:
@wckzhang Most the of the above functions and structures have htod and dtoh variants. Code duplication can be reduced by introducing htod, dtoh enums and passing that as argument to functions or having that as loop index. What do you think?

Reply:
Yeah it can, I'm trying to (it's not working too well) make as few changes as possible. This PR is big as is, and I'd like it to get merged before we do too many changes.

Reply:
In the interest in getting this PR moved, I'm not going to make these changes.